### PR TITLE
Adapt MSVC C++ mode handling to VS15.5

### PR DESCRIPTION
**Pull request by hkaiser**

### CODE REVIEW

File: cmake/HPX_DetectCppDialect.cmake [hunk 1]
```diff
@@ -116,6 +112,9 @@ macro(hpx_detect_cpp_dialect)
       hpx_info("C++ mode assumed: C++11")
     endif()
 
+    hpx_add_config_cond_define(_HAS_AUTO_PTR_ETC 1)
+    hpx_add_config_cond_define(_SILENCE_ALL_CXX17_DEPRECATION_WARNINGS)
```
Comment (sithhell) [inline cmake/HPX_DetectCppDialect.cmake]: Shouldn't this go in the `HPX_WITH_CXX17` only?

File: cmake/HPX_DetectCppDialect.cmake [hunk 2]
```diff
@@ -89,24 +89,20 @@ macro(hpx_detect_cpp_dialect)
 
     # enable enforcing a particular C++ mode
     if(HPX_WITH_CXX17)
-      set(CXX_FLAG -std:c++latest)
-      hpx_add_config_cond_define(_HAS_AUTO_PTR_ETC 1)
+      set(CXX_FLAG -std:c++17)
       hpx_info("C++ mode enforced: C++17")
     elseif(HPX_WITH_CXX1Z)
       set(CXX_FLAG -std:c++latest)
```
Comment (sithhell) [inline cmake/HPX_DetectCppDialect.cmake]: C++1z essentially is C++17, right?

Comment (hkaiser) [reply cmake/HPX_DetectCppDialect.cmake]: Well, not quite - doesn't really change anything, however.

File: cmake/HPX_DetectCppDialect.cmake (hunk 1 above)
Comment (hkaiser) [reply cmake/HPX_DetectCppDialect.cmake]: Yes, we can do that.

File: cmake/HPX_DetectCppDialect.cmake (hunk 2 above)
Comment (sithhell) [reply cmake/HPX_DetectCppDialect.cmake]: my point is, when the next version of MSVC is released, `-std:c++latest` might refer to C++2a, right?

Comment (hkaiser) [reply cmake/HPX_DetectCppDialect.cmake]: Who knows.

File: cmake/HPX_DetectCppDialect.cmake [hunk 3]
```diff
@@ -89,24 +89,21 @@ macro(hpx_detect_cpp_dialect)
 
     # enable enforcing a particular C++ mode
     if(HPX_WITH_CXX17)
-      set(CXX_FLAG -std:c++latest)
-      hpx_add_config_cond_define(_HAS_AUTO_PTR_ETC 1)
+      set(CXX_FLAG -std:c++17)
       hpx_info("C++ mode enforced: C++17")
+      hpx_add_config_cond_define(_SILENCE_ALL_CXX17_DEPRECATION_WARNINGS)
     elseif(HPX_WITH_CXX1Z)
       set(CXX_FLAG -std:c++latest)
```
Comment (sithhell) [inline cmake/HPX_DetectCppDialect.cmake]: I am still unhappy with this. Isn't C++17 the latest and not C++1z?

**Pull request merged**